### PR TITLE
misc: add the s5_trigger.sh script

### PR DESCRIPTION
**Pull request by peterfang**
misc: add the s5_trigger.sh script

Tracked-On: #5411
Signed-off-by: Peter Fang <peter.fang@intel.com>

### CODE REVIEW

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: misc/life_mngr/s5_trigger.sh [hunk 1]
```diff
@@ -0,0 +1,214 @@
+#! /bin/bash
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:1]: Not sure if strictly needed but we usually include the copyright and license information in our scripts.

File: misc/life_mngr/s5_trigger.sh [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	read -r -d $'\0' ret_msg < /dev/ttyS1
+}
+
+power_off_post_vms() {
+	vm_list=$(acrnctl list)
+	echo $vm_list
+
+	array=($vm_list)
+	num=${#array[@]}
+	echo "vm array size:" $num
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:50]:
```suggestion
	echo "Number of VMs:" $num
```

Would this be appropriate?

File: misc/life_mngr/s5_trigger.sh [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	if [ $num -gt $MAX_ARRAY ]; then
+		echo "not normal for acrnctl, or no VMs!"
+		return 0
+	fi
+
+	#shut down post-launched VMs
+	for ((i=0; i<$num; i+=2))
+	do
+		if [ ${array[$i+1]} == $started ]; then
+			echo "shutting down: " ${array[$i]}
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:61]:
```suggestion
			echo "Shutting down: " ${array[$i]}
```

File: misc/life_mngr/s5_trigger.sh (hunk 1 above)
Comment (gvancuts) [reply misc/life_mngr/s5_trigger.sh:1]: Where do all these `echo` messages end up? This script is called directly by `life_mngr` so I'm wondering how these will be exposed to the user?

File: misc/life_mngr/s5_trigger.sh [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+check_post_vms_status() {
+	#check post-launched VM status for some time
+	check_times=5
+	while [[ $check_times > 0 ]]
+	do
+		vm_list=$(acrnctl list)
+		array=($vm_list)
+		num=${#array[@]}
+		echo "vm status: " $vm_list
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:78]:
```suggestion
		echo "VM status: " $vm_list
```

File: misc/life_mngr/s5_trigger.sh [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+power_off_post_vms() {
+	vm_list=$(acrnctl list)
+	echo $vm_list
+
+	array=($vm_list)
+	num=${#array[@]}
+	echo "vm array size:" $num
+
+	if [ $num -gt $MAX_ARRAY ]; then
+		echo "not normal for acrnctl, or no VMs!"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:53]:
```suggestion
		echo "There are no VMs running or acrnctl encountered an internal error."
```

File: misc/life_mngr/s5_trigger.sh [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+			sleep 5s
+			let check_times--
+			echo "check time:" $check_times
+		else
+			echo "vm status: " $vm_list
+			break;
+		fi
+
+		if [ $check_times -eq 0 ]; then
+			echo "timed out waiting for vms..."
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:107]:
```suggestion
			echo "Timed out waiting for VMs..."
```

For VMs to what? To shutdown?

File: misc/life_mngr/s5_trigger.sh [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+				break;
+			fi
+		done
+
+		if [ $flag -eq 1 ]; then
+			sleep 5s
+			let check_times--
+			echo "check time:" $check_times
+		else
+			echo "vm status: " $vm_list
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:102]:
```suggestion
			echo "VM status: " $vm_list
```

File: misc/life_mngr/s5_trigger.sh [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+check_post_vms_alive() {
+	#check if there is any post-launched VM alive, and return if there is
+	vm_list=$(acrnctl list)
+	array=($vm_list)
+	num=${#array[@]}
+
+	for ((i=0; i<$num; i+=2))
+	do
+		if [ ${array[$i+1]} == $started ]; then
+			echo $vm_list " vm alive!"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:128]:
```suggestion
			echo $vm_list " VM alive!"
```

File: misc/life_mngr/s5_trigger.sh [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+			return 1
+		fi
+	done
+
+	echo "no vm alive: " $vm_list
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:133]:
```suggestion
	echo "No VM alive: " $vm_list
```

File: misc/life_mngr/s5_trigger.sh [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+	return 0
+}
+
+if [ "$1" = "$SOS" ]; then
+	try_times=2
+
+	#shut down post-launched VMs
+	while [[ $try_times -gt 0 ]]
+	do
+		echo "check post-launched vms alive status..."
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:143]:
```suggestion
		echo "check post-launched VMs status..."
```

File: misc/life_mngr/s5_trigger.sh [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+		check_post_vms_alive
+		if [ $? -eq 0 ]; then
+			try_times=1
+			break
+		fi
+
+		echo "power off vms..."
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:150]:
```suggestion
		echo "Powering off VMs..."
```

File: misc/life_mngr/s5_trigger.sh [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+		power_off_post_vms
+		if [ $? -eq 0 ]; then
+			break
+		fi
+
+		echo "check vms status..."
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:156]:
```suggestion
		echo "Checking VMs status..."
```

File: misc/life_mngr/s5_trigger.sh [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+		check_post_vms_status
+		if [ $? -eq 1 ]; then
+			break
+		fi
+
+		let try_times--
+	done
+
+	if [ $try_times -eq 0 ]; then
+		echo "S5 failed! please check!"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:166]:
```suggestion
		echo "S5 failed!"
```

File: misc/life_mngr/s5_trigger.sh [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+	#send shutdown message to the pre-launched VM
+	send_pre_message $SHUTDOWN
+	#read ack message from the pre-launched VM
+	read_pre_message
+	echo "ret_msg: $ret_msg"
+
+	#check the ack message
+	if [ "$ret_msg" = "$ACKED" ]; then
+		echo "received ack message"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:180]:
```suggestion
		echo "Received ACK message"
```

File: misc/life_mngr/s5_trigger.sh [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+	else
+		echo "can not receive ack message from pre-launched OS, pls check!"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:182]:
```suggestion
		echo "Could not receive ACK message from pre-launched User VM, please check!"
```

File: misc/life_mngr/s5_trigger.sh [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+		exit
+	fi
+
+	#all pre-launched and post-launched VMs have shut down
+	echo "shutting down SOS itself..."
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:187]:
```suggestion
	echo "Shutting down the Service VM itself..."
```

File: misc/life_mngr/s5_trigger.sh [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+	sleep 3s
+	poweroff
+
+elif [ "$1" = "$UOS" ]; then
+	echo "trying to open socket..."
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:192]:
```suggestion
	echo "Trying to open socket..."
```

File: misc/life_mngr/s5_trigger.sh [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+	if ! exec 6<>/dev/tcp/$IP_ADDR/$SOCKET_PORT
+	then
+		echo "open socked failed"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:195]:
```suggestion
		echo "Failed to open socket"
```

File: misc/life_mngr/s5_trigger.sh [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+		exit 1
+	fi
+	echo "socket is open"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:198]:
```suggestion
	echo "Socket opened"
```

File: misc/life_mngr/s5_trigger.sh [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+
+	#send shutdown message
+	send_message $SHUTDOWN
+	#read ack message
+	read_message
+	echo "ret_msg: $ret_msg"
+
+	#check the ack message
+	if [ "$ret_msg" = "$ACKED" ]; then
+		echo "received ack message"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:208]:
```suggestion
		echo "Received ACK message"
```

File: misc/life_mngr/s5_trigger.sh [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+	else
+		echo "can not receive ack message, pls check!"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:210]:
```suggestion
		echo "Could not receive ACK message, please check!"
```

File: misc/life_mngr/s5_trigger.sh [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+	fi
+else
+	echo "error! pls use: ./s5_trigger.sh sos  OR ./s5_trigger.sh uos"
```
Comment (gvancuts) [inline misc/life_mngr/s5_trigger.sh:213]:
```suggestion
	echo "Error! Please use: ./s5_trigger.sh sos  OR ./s5_trigger.sh uos"
```

**Review by gvancuts — CHANGES_REQUESTED** [review]
We should also make sure this is installed by `make install`. Can you add this?

File: misc/life_mngr/s5_trigger.sh (hunk 1 above)
Comment (peterfang) [reply misc/life_mngr/s5_trigger.sh:1]: > Where do all these `echo` messages end up? This script is called directly by `life_mngr` so I'm wondering how these will be exposed to the user?

There are numerous use cases for the script. The user can execute the script directly and they can see the output in their shell. If it's called inside life_mngr, then they don't see the `echo` messages.

Comment (gvancuts) [reply misc/life_mngr/s5_trigger.sh:1]: ok, thanks for the clarification

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Review by ywan170 — APPROVED** [review]

**Review by terryzouhao — APPROVED** [review]

**Comment by wenlingz** [comment]
@peterfang  please don't install HOME path, modify this para later
install -p -m 0755 s5_trigger.sh $(HOME)

**Comment by peterfang** [comment]
> @peterfang please don't install HOME path, modify this para later
> install -p -m 0755 s5_trigger.sh $(HOME)

Unfortunately, life_mngr is hard-coded to run s5_trigger.sh in $(HOME) (has been this way since the beginning).

**Comment by peterfang** [comment]
> We should also make sure this is installed by `make install`. Can you add this?

It was reported that we can't install to $(HOME) (hard-coded location in life_mngr) so we've agreed to not do this right now.

**Comment by HuiHuangShi** [comment]
No new violations to the coding guideline detected.<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by gvancuts** [comment]
> It was reported that we can't install to $(HOME) (hard-coded location in life_mngr) so we've agreed to not do this right now.

OK. Can you add a note to the documentation explaining that the user needs to take the scripts and manually install it in the right location?

Is the plan moving forward to update the code to change that?

I also noticed that `life_mngr.service` is not installed either as part of the `make install` steps. Ideally, we'd install it **and** enable the service by default.

**Comment by peterfang** [comment]
> > It was reported that we can't install to $(HOME) (hard-coded location in life_mngr) so we've agreed to not do this right now.
> 
> OK. Can you add a note to the documentation explaining that the user needs to take the scripts and manually install it in the right location?
> 
> Is the plan moving forward to update the code to change that?
> 
> I also noticed that `life_mngr.service` is not installed either as part of the `make install` steps. Ideally, we'd install it **and** enable the service by default.

Sure, I can update my doc PR to give a bit more explanation. Currently, there is no plan to change that.
`life_mngr.service` is not a necessary service for ACRN and it needs to be installed in guest OS's as well. We advise anyone who has a need for graceful (coordinated) system shutdown to follow the `Enable S5 in ACRN` documentation.

Please use this PR for further discussions regarding documentation: https://github.com/projectacrn/acrn-hypervisor/pull/5541

**Pull request merged**